### PR TITLE
Move demo to separate distribution

### LINKED ISSUE #904: Extract Demo App into an independently deployable package
Currently the demo application is tied to the deployment of TraitsUI. Since TraitsUI is a library heavily depended on by other packages, it needs to be stable and should have fewer external dependencies. On the other hand, the demo application, um, is an application which can be developed and deployed more frequently and maybe with more external dependencies. The latter is more achievable if it is decoupled from TraitsUI.

### PR DESCRIPTION
**Pull request by kitchoi**
Closes #904 

This PR
- [x] Move the content related to the demo application to a new distribution. (There are a handful flake8 fixes so that the new distribution is covered by flake8)
- [x] Copy the original `traitsui/extras/demo.py` to `traitsui/extras/_demo_legacy.py` and add an entry point back to `traitsui/extras/demo.py` so that the `traitsui.extras.demo.demo` is still importable by packages (e.g. the `tutor.py` in `chaco/examples`). 
- [x] Add README under the `ets-demo` folder
- [x] Add `setup.py` under the `ets-demo` folder (adapted from the combination of traitsui's and traits's `setup.py`)
- [x] Add `etstools.py` under `ets-demo` folder (adapted from traitsui's `etstools.py`)
- [x] Update Travis and Appveyor setup to run tests in `ets-demo`
- ~Hard-code traitsui example path (for now) so that the demo application shows something. In the future, data should be contributed via entry points.~ Not possible at the moment, we should just do entry points.
- [x] Confirm the names for the new folder / distribution / package. <--- Need discussion as part of the review

### CODE REVIEW

**Comment by kitchoi** [comment]
> Hard-code traitsui example path (for now) so that the demo application shows something. In the future, data should be contributed via entry points.

I can't hard-code this because the traitsui's examples folder is only available from the source distribution, not everyone install from source. e.g. `pip install traitsui` won't include in the example files.

File: appveyor-run.cmd [hunk 1]
```diff
@@ -7,14 +7,15 @@
...
+PUSHD ets-demo
+CALL edm run -- python etstool.py !operation! --runtime=!runtime! --toolkit=!toolkit! || GOTO error
+POPD
 
-FOR %%x IN (%*) DO (
-SET /A counter=!counter! + 1
-IF !counter! EQU 1 SET operation=%%x
-IF !counter! EQU 2 SET runtime=%%x
-IF !counter! GTR 2 CALL edm run -- python etstool.py !operation! --runtime=!runtime! --toolkit=%%x || GOTO error
-)
```
Comment (kitchoi) [inline appveyor-run.cmd:17]: Drive-by structural changes to the batch script. The original script took me a while to understand. I don't know what called for the initial indirection, but whatever it was trying to workaround does not seem to apply now.

File: ets-demo/etsdemo/tests/test_app.py [hunk 2]
```diff
@@ -144,3 +148,87 @@ def test_init_dict_used_by_children(self):
             example.run_code()
             self.assertIn("CONSTANT", example.locals)
             self.assertIn("HasTraits", example.locals)
+
+
+class TestParseSource(unittest.TestCase):
```
Comment (kitchoi) [inline ets-demo/etsdemo/tests/test_app.py:153]: This test is copied from `traitsui/tests/test_extras`. We can reorganize functions and modules later.

File: ets-demo/setup.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+# Paths to the autogenerated version file and the Git directory.
+HERE = os.path.abspath(os.path.dirname(__file__))
+VERSION_FILE = os.path.join(HERE, "etsdemo", "version.py")
+# etsdemo is sharing the same Git directory with traitsui
+GIT_DIRECTORY = os.path.join(HERE, "..", ".git")
+
+# Template for the autogenerated version file.
+VERSION_FILE_TEMPLATE = '''\
+# (C) Copyright 2020-2020 {company}, Inc., Austin, TX
```
Comment (kitchoi) [inline ets-demo/setup.py:43]: A lot of this copied from traits. I did change the copyright year start to 2020.

**Comment by kitchoi** [comment]
Ready for review. 

Note that I have tried to minimize any functional change. In fact, there are no functional changes apart from code being moved and minor flake8 fixes.
The majority of actual new code are the `setup.py`, `etstools.py`, but a large part of them are replicates of other ETS packages (for consistencies and not to reinvent the wheel).

I wrote the README, named the package, and updated the CI setup, so these could do with a bit more careful review.

File: traitsui/extras/_demo_legacy.py [hunk 4]
```diff
@@ -0,0 +1,1187 @@
...
+#  Date:   09/15/2005
+#
+# -------------------------------------------------------------------------
+
+""" A Traits UI demo that borrows heavily from the design of the wxPython demo.
+
+This copy of the demo application will only be maintained until ``ets-demo``
+is released and distributed. This allows other ETS packages to continue to be
+able to import the demo application until ``ets-demo`` is released.
+"""
```
Comment (kitchoi) [inline traitsui/extras/_demo_legacy.py:23]: The rest of this file was identical to the removed `traitsui/extras/demo.py`, except for this paragraph.
I should open a new issue for removing this legacy file eventually. Will do that now.

**Comment by kitchoi** [comment]
Travis CI build has completed and passed (i.e. excluding the existing wx failures), but the webhook response has not reached GitHub.

**Comment by kitchoi** [comment]
> One big thing is that right now a user script is needed to launch the demo from the distribution. Maybe a script similar to traitsui/examples/tutorials/tutor.py should be added to distribution that allows users to launch the application with an arbitrary demo folder?

Yep there should be a new main function for launching the GUI, to be exposed in the entry point, but that will be new code. It will also be needed as part of #908.
For the purpose of this PR, I wanted to restrict the scope to only moving existing code to a new distribution and adding all the necessary build related scripts. In other words, I intend to add that "script" in a separate PR.

File: ets-demo/etstool.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+    python etstool.py cleanup --runtime=... --toolkit=...
+
+If you make changes you will either need to remove and re-install the
+environment or manually update the environment using ``edm``, as
+the install performs a ``python setup.py install`` rather than a ``develop``,
+so changes in your code will not be automatically mirrored in the test
+environment.  You can update with a command like::
+
+    edm run --environment ... -- python setup.py install
```
[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/etstool.py:45]: I think it is needed in `traits` because of the occasional need to recompile the C code, which I often just do manually.
With the `--editable` opiton in `install`, there is never a need for me to call the `update` command in any of these `etstool.py`. I often just reinstall for a clean slate.  I might be alone in that, not sure. Should I add it anyway?

[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/etstool.py:45]: I guess I will let someone else who needs to add it.

File: ets-demo/etstool.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    "docutils",
+    "flake8",
+    "traits",
+    "traitsui",
+    "pyface",
+    "pip",
+}
+
+extra_dependencies = {
+    'pyside': {'pyside'},
```
[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/etstool.py:114]: Well spotted!  I was trying to avoid interfering with #821 and has clearly failed :D

File: ets-demo/etstool.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+import subprocess
+import sys
+from shutil import rmtree, copy as copyfile
+from tempfile import mkdtemp
+from contextlib import contextmanager
+
+import click
+
+supported_combinations = {
+    '3.5': {'pyside2', 'pyqt', 'pyqt5', 'null'},
```
[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/etstool.py:90]: Yep I can drop 3.5 in `etstools.py` here as we are not testing for Python 3.5 on CI and I bet no users of this `etstool.py` cares. I will keep the one in `setup.py` to be consistent with other projects though!

File: ets-demo/etstool.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ievacerny) [reply ets-demo/etstool.py:45]: Personally I don't use `update` either, didn't even know it existed until I compared the docstrings. It seems that `etstool.py` in Traits is more often updated, so I thought maybe this is a nice new thing that should be brought into other `etstool.py` scripts too. But I don't have a strong opinion about this, so I'm going to leave it up to you.

File: ets-demo/etstool.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ievacerny) [reply ets-demo/etstool.py:117]: Sorry if I wasn't clear - `pyside` should be removed from extra dependencies because it's not supported. Support for it shouldn't be added.

[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/etstool.py:117]: But when we proceed with bringing PyQt4 and PySide back to CI for TraitsUI, this distribution will be affected because this distribution is still tied with TraitsUI on CI. There is a potential for PR interactions when we modify CI build configurations. So I think the safest thing to do is to let this distribution support all the backends we are going to support once #815 is fixed.

[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/etstool.py:117]: Sorry, been referencing the PR, I should be referring to #821

Now I looked at again, `"pyside"` was not in `supported_combinations` when #815 removed support for PySide, is that what you meant?

[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/etstool.py:117]: Because it does not support Python 3, of course. Yep, it should be removed.

File: ets-demo/setup.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+            Intended Audience :: Developers
+            Intended Audience :: Science/Research
+            License :: OSI Approved :: BSD License
+            Operating System :: MacOS
+            Operating System :: Microsoft :: Windows
+            Operating System :: OS Independent
+            Operating System :: POSIX
+            Operating System :: Unix
+            Programming Language :: Python
+            Programming Language :: Python :: 3.5
```
[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/setup.py:273]: I hope to keep this for now to be in-sync with other ETS projects.

**Comment by kitchoi** [comment]
Thank you @ievacerny 
Ready for re-review (I hope I addressed everything).

File: ets-demo/etsdemo/app.py [hunk 9]
```diff
@@ -0,0 +1,1186 @@
...
+#
+#  This software is provided without warranty under the terms of the BSD
+#  license included in LICENSE.txt and may be redistributed only
+#  under the conditions described in the aforementioned license.  The license
+#  is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+#  Thanks for using Enthought open source!
+#
+#  Author: David C. Morrill
+#  Date:   09/15/2005
```
[the remark this answers is not in the document]
Comment (ievacerny) [reply ets-demo/etsdemo/app.py]: Copyright at the top of `etstool.py` still needs updating

[the remark this answers is not in the document]
Comment (kitchoi) [reply ets-demo/etsdemo/app.py]: Oops 🤦  Thanks!

**Comment by kitchoi** [comment]
@corranwebster Would you be able to take a look at this please?

File: ets-demo/.gitignore [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# file types to ignore
+*.pyc
+*.pyd
+*.so
+*.enamlc
```
Comment (rahulporuri) [inline ets-demo/.gitignore:5]: i think we can finally ditch the `.enamlc` files from the gitignore.

Comment (rahulporuri) [reply ets-demo/.gitignore:5]: nit-picky : possibly even the `.so` files. i don't think there are any compiled modules in traitsui, are there?

File: ets-demo/setup.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+
+# Paths to the autogenerated version file and the Git directory.
+HERE = os.path.abspath(os.path.dirname(__file__))
+VERSION_FILE = os.path.join(HERE, "etsdemo", "version.py")
+# etsdemo is sharing the same Git directory with traitsui
+GIT_DIRECTORY = os.path.join(HERE, "..", ".git")
+
+# Template for the autogenerated version file.
+VERSION_FILE_TEMPLATE = '''\
+# (C) Copyright 2020-2020 Enthought, Inc., Austin, TX
```
Comment (rahulporuri) [inline ets-demo/setup.py:43]: i think it just needs to be `2020`. not 2020-2020.

File: ets-demo/.gitignore (hunk 10 above)
Comment (kitchoi) [reply ets-demo/.gitignore:5]: I don't think they are needed, but at the same time if they are here we also don't want to track them. These are probably just copied from everywhere (I copied from traitsui) without much curating so that they can be copied again to other places, again without much curating...

Comment (kitchoi) [reply ets-demo/.gitignore]: They are removed now nonetheless!

**Comment by kitchoi** [comment]
Thank you both. Merging...

File: appveyor-run.cmd (hunk 1 above)
Comment (kitchoi) [reply appveyor-run.cmd:17]: Post-mortem: turns out the loop here is such that we can run this batch script like this:
```
appveyor-run.cmd install "3.6" null pyqt pyqt4 wx
```
The change here means the rest of the toolkit after the first one is ignore.
#971 sets the toolkit argument to a single value in the build configuration.
